teamcity: failed tests on master: testrace/TestSnapshotAfterTruncation, testrace/TestSnapshotAfterTruncation/differentTerm

The following tests appear to have failed:

[#351033](https://teamcity.cockroachdb.com/viewLog.html?buildId=351033):

```
--- FAIL: testrace/TestSnapshotAfterTruncation (1.560s)
null
--- FAIL: testrace/TestSnapshotAfterTruncation/differentTerm (1.120s)
client_raft_test.go:792: 2: expected last index 25, but found 22
------- Stdout: -------
I170913 14:25:11.455832 41754 storage/store.go:1208  [s1] [n1,s1]: failed initial metrics computation: [n1,s1]: system config not yet available
I170913 14:25:11.456149 41754 gossip/gossip.go:311  [n1] NodeDescriptor set to node_id:1 address:<network_field:"tcp" address_field:"127.0.0.1:39922" > attrs:<> locality:<> ServerVersion:<major_val:0 minor_val:0 patch:0 unstable:0 > 
W170913 14:25:11.489909 41754 gossip/gossip.go:1225  [n2] no incoming or outgoing connections
I170913 14:25:11.494372 41908 gossip/client.go:129  [n2] started gossip client to 127.0.0.1:39922
I170913 14:25:11.521452 41754 gossip/gossip.go:311  [n2] NodeDescriptor set to node_id:2 address:<network_field:"tcp" address_field:"127.0.0.1:37932" > attrs:<> locality:<> ServerVersion:<major_val:0 minor_val:0 patch:0 unstable:0 > 
W170913 14:25:11.535771 41754 gossip/gossip.go:1225  [n3] no incoming or outgoing connections
I170913 14:25:11.547009 41919 gossip/client.go:129  [n3] started gossip client to 127.0.0.1:39922
I170913 14:25:11.550089 41754 storage/store.go:1208  [s3] [n3,s3]: failed initial metrics computation: [n3,s3]: system config not yet available
I170913 14:25:11.552587 41754 gossip/gossip.go:311  [n3] NodeDescriptor set to node_id:3 address:<network_field:"tcp" address_field:"127.0.0.1:38967" > attrs:<> locality:<> ServerVersion:<major_val:0 minor_val:0 patch:0 unstable:0 > 
I170913 14:25:11.585251 41899 storage/replica_raftstorage.go:517  [s1,r1/1:/M{in-ax}] generated preemptive snapshot 6023e97a at index 16
I170913 14:25:11.591540 41899 storage/store.go:3502  [s1,r1/1:/M{in-ax}] streamed snapshot to (n2,s2):?: kv pairs: 35, log entries: 6, rate-limit: 2.0 MiB/sec, 5ms
I170913 14:25:11.592623 41902 storage/replica_raftstorage.go:725  [s2,r1/?:{-}] applying preemptive snapshot at index 16 (id=6023e97a, encoded size=6096, 1 rocksdb batches, 6 log entries)
I170913 14:25:11.593875 41902 storage/replica_raftstorage.go:733  [s2,r1/?:/M{in-ax}] applied preemptive snapshot in 1ms [clear=0ms batch=0ms entries=0ms commit=0ms]
I170913 14:25:11.596258 41899 storage/replica_command.go:3647  [s1,r1/1:/M{in-ax}] change replicas (ADD_REPLICA (n2,s2):2): read existing descriptor r1:/M{in-ax} [(n1,s1):1, next=2]
I170913 14:25:11.607763 42136 storage/replica.go:3032  [s1,r1/1:/M{in-ax}] proposing ADD_REPLICA((n2,s2):2): updated=[(n1,s1):1 (n2,s2):2] next=3
I170913 14:25:11.615160 42178 storage/replica_raftstorage.go:517  [s1,r1/1:/M{in-ax}] generated preemptive snapshot b0b640a6 at index 18
I170913 14:25:11.618694 42178 storage/store.go:3502  [s1,r1/1:/M{in-ax}] streamed snapshot to (n3,s3):?: kv pairs: 38, log entries: 8, rate-limit: 2.0 MiB/sec, 3ms
I170913 14:25:11.619932 42017 storage/replica_raftstorage.go:725  [s3,r1/?:{-}] applying preemptive snapshot at index 18 (id=b0b640a6, encoded size=7643, 1 rocksdb batches, 8 log entries)
I170913 14:25:11.621441 42017 storage/replica_raftstorage.go:733  [s3,r1/?:/M{in-ax}] applied preemptive snapshot in 1ms [clear=0ms batch=0ms entries=1ms commit=0ms]
I170913 14:25:11.625201 42178 storage/replica_command.go:3647  [s1,r1/1:/M{in-ax}] change replicas (ADD_REPLICA (n3,s3):3): read existing descriptor r1:/M{in-ax} [(n1,s1):1, (n2,s2):2, next=3]
I170913 14:25:11.630220 42166 storage/raft_transport.go:455  raft transport stream to node 1 established
I170913 14:25:11.638208 42168 storage/replica.go:3032  [s1,r1/1:/M{in-ax}] proposing ADD_REPLICA((n3,s3):3): updated=[(n1,s1):1 (n2,s2):2 (n3,s3):3] next=4
W170913 14:25:11.798608 42195 storage/raft_transport.go:282  unable to accept Raft message from (n1,s1):1: no handler registered for (n2,s2):2
W170913 14:25:11.799527 42142 storage/store.go:3238  [s1] raft error: node 2 claims to not contain store 2 for replica (n2,s2):2: store 2 was not found
W170913 14:25:11.804211 42182 storage/raft_transport.go:461  raft transport stream to node 2 failed: store 2 was not found
I170913 14:25:11.806657 42185 storage/raft_transport.go:455  raft transport stream to node 2 established
W170913 14:25:11.810293 42212 storage/raft_transport.go:282  unable to accept Raft message from (n1,s1):1: no handler registered for (n2,s2):2
W170913 14:25:11.811854 42187 storage/raft_transport.go:495  no handler found for store 1 in response range_id:1 from_replica:<node_id:2 store_id:2 replica_id:2 > to_replica:<node_id:1 store_id:1 replica_id:1 > union:<error:<message:"store 2 was not found" transaction_restart:NONE origin_node:0 detail:<store_not_found:<store_id:2 > > now:<wall_time:0 logical:0 > > > 
W170913 14:25:11.812253 42197 storage/raft_transport.go:282  unable to accept Raft message from (n3,s3):3: no handler registered for (n1,s1):1
W170913 14:25:11.812989 41872 storage/store.go:3238  [s3] raft error: node 1 claims to not contain store 1 for replica (n1,s1):1: store 1 was not found
W170913 14:25:11.813150 42166 storage/raft_transport.go:461  raft transport stream to node 1 failed: store 1 was not found
I170913 14:25:12.070399 42176 storage/client_test.go:1194  [hb] test clock advanced to: 1.800000125,0
W170913 14:25:12.071891 42311 storage/node_liveness.go:358  [hb] failed node liveness heartbeat: context deadline exceeded
I170913 14:25:12.098297 42224 storage/client_test.go:1194  [hb] test clock advanced to: 3.600000127,0
W170913 14:25:12.099665 42397 storage/node_liveness.go:358  [hb] failed node liveness heartbeat: context deadline exceeded
I170913 14:25:12.129927 42484 storage/client_test.go:1194  [hb] test clock advanced to: 5.400000129,0
W170913 14:25:12.131173 42471 storage/node_liveness.go:358  [hb] failed node liveness heartbeat: context deadline exceeded
I170913 14:25:12.195758 42515 storage/raft_transport.go:455  raft transport stream to node 1 established
E170913 14:25:12.454099 42238 storage/replica.go:4056  [s1,r1/1:/M{in-ax}] unable to add replica to Raft repair queue: queue disabled
I170913 14:25:12.459234 41754 storage/replica_raftstorage.go:517  [raftsnapshot,s1,r1/1:/M{in-ax}] generated Raft snapshot 594479cb at index 22
I170913 14:25:12.462094 41754 storage/store.go:3502  [raftsnapshot,s1,r1/1:/M{in-ax}] streamed snapshot to (n2,s2):2: kv pairs: 42, log entries: 1, rate-limit: 8.0 MiB/sec, 2ms
I170913 14:25:12.462685 42520 storage/replica_raftstorage.go:725  [s2,r1/2:/M{in-ax}] applying Raft snapshot at index 22 (id=594479cb, encoded size=4706, 1 rocksdb batches, 1 log entries)
I170913 14:25:12.463758 42520 storage/replica_raftstorage.go:733  [s2,r1/2:/M{in-ax}] applied Raft snapshot in 1ms [clear=0ms batch=0ms entries=0ms commit=0ms]
I170913 14:25:12.493840 42486 util/stop/stopper.go:476  quiescing; tasks left:
3      storage.Replica: acquiring lease to gossip
3      [async] storage.multiTestContextKVTransport: calling next replica
I170913 14:25:12.494457 42486 util/stop/stopper.go:476  quiescing; tasks left:
3      [async] storage.multiTestContextKVTransport: calling next replica
2      storage.Replica: acquiring lease to gossip
I170913 14:25:12.495145 42486 util/stop/stopper.go:476  quiescing; tasks left:
3      [async] storage.multiTestContextKVTransport: calling next replica
1      storage.Replica: acquiring lease to gossip
I170913 14:25:12.503727 42486 util/stop/stopper.go:476  quiescing; tasks left:
3      [async] storage.multiTestContextKVTransport: calling next replica
W170913 14:25:12.504247 42518 storage/replica.go:2686  [hb,s1,r1/1:/M{in-ax}] shutdown cancellation after 0.1s of attempting command [txn: 4aa1632f], BeginTransaction [/System/NodeLiveness/2,/Min), ConditionalPut [/System/NodeLiveness/2,/Min), EndTransaction [/System/NodeLiveness/2,/Min)
I170913 14:25:12.505312 42486 util/stop/stopper.go:476  quiescing; tasks left:
2      [async] storage.multiTestContextKVTransport: calling next replica
W170913 14:25:12.505633 42471 storage/node_liveness.go:358  [hb] failed node liveness heartbeat: node unavailable; try another peer
W170913 14:25:12.506068 42353 storage/replica.go:2686  [hb,s1,r1/1:/M{in-ax}] shutdown cancellation after 0.2s of attempting command [txn: fb2357aa], BeginTransaction [/System/NodeLiveness/1,/Min), ConditionalPut [/System/NodeLiveness/1,/Min), EndTransaction [/System/NodeLiveness/1,/Min)
W170913 14:25:12.507604 42311 storage/node_liveness.go:358  [hb] failed node liveness heartbeat: node unavailable; try another peer
I170913 14:25:12.507863 42486 util/stop/stopper.go:476  quiescing; tasks left:
1      [async] storage.multiTestContextKVTransport: calling next replica
W170913 14:25:12.508447 42530 storage/replica.go:2686  [hb,s1,r1/1:/M{in-ax}] shutdown cancellation after 0.2s of attempting command [txn: 553fc543], BeginTransaction [/System/NodeLiveness/3,/Min), ConditionalPut [/System/NodeLiveness/3,/Min), EndTransaction [/System/NodeLiveness/3,/Min)
W170913 14:25:12.510122 42397 storage/node_liveness.go:358  [hb] failed node liveness heartbeat: node unavailable; try another peer
I170913 14:25:12.534527 41313 vendor/google.golang.org/grpc/grpclog/grpclog.go:75  transport: http2Server.HandleStreams failed to read frame: read tcp 127.0.0.1:37932->127.0.0.1:48820: use of closed network connection
I170913 14:25:12.534588 41680 vendor/google.golang.org/grpc/grpclog/grpclog.go:60  transport: http2Client.notifyError got notified that the client transport was broken EOF.
I170913 14:25:12.534825 41633 vendor/google.golang.org/grpc/grpclog/grpclog.go:75  transport: http2Server.HandleStreams failed to read frame: read tcp 127.0.0.1:39922->127.0.0.1:34265: use of closed network connection
I170913 14:25:12.535024 40400 vendor/google.golang.org/grpc/grpclog/grpclog.go:60  transport: http2Client.notifyError got notified that the client transport was broken EOF.
I170913 14:25:12.535869 41866 vendor/google.golang.org/grpc/grpclog/grpclog.go:75  transport: http2Server.HandleStreams failed to read frame: read tcp 127.0.0.1:38967->127.0.0.1:41651: use of closed network connection
I170913 14:25:12.536013 42036 vendor/google.golang.org/grpc/grpclog/grpclog.go:60  transport: http2Client.notifyError got notified that the client transport was broken EOF.
I170913 14:25:12.536155 41779 vendor/google.golang.org/grpc/grpclog/grpclog.go:75  grpc: addrConn.resetTransport failed to create client transport: connection error: desc = "transport: Error while dialing dial tcp 127.0.0.1:39922: getsockopt: connection refused"; Reconnecting to {127.0.0.1:39922 <nil>}
I170913 14:25:12.536590 41923 vendor/google.golang.org/grpc/grpclog/grpclog.go:75  grpc: addrConn.resetTransport failed to create client transport: connection error: desc = "transport: Error while dialing dial tcp 127.0.0.1:37932: getsockopt: connection refused"; Reconnecting to {127.0.0.1:37932 <nil>}
I170913 14:25:12.536955 42039 vendor/google.golang.org/grpc/grpclog/grpclog.go:75  grpc: addrConn.resetTransport failed to create client transport: connection error: desc = "transport: Error while dialing dial tcp 127.0.0.1:38967: getsockopt: connection refused"; Reconnecting to {127.0.0.1:38967 <nil>}
I170913 14:25:12.537409 41923 vendor/google.golang.org/grpc/grpclog/grpclog.go:75  grpc: addrConn.transportMonitor exits due to: context canceled
I170913 14:25:12.537518 42039 vendor/google.golang.org/grpc/grpclog/grpclog.go:75  grpc: addrConn.transportMonitor exits due to: context canceled
I170913 14:25:12.537776 41779 vendor/google.golang.org/grpc/grpclog/grpclog.go:75  grpc: addrConn.transportMonitor exits due to: context canceled
W170913 14:25:12.538423 42198 storage/raft_transport.go:461  raft transport stream to node 3 failed: rpc error: code = Unavailable desc = transport is closing

```
Please assign, take a look and update the issue accordingly.

--- a/pkg/storage/client_raft_test.go
+++ b/pkg/storage/client_raft_test.go
@@ -772,29 +772,32 @@ func TestSnapshotAfterTruncation(t *testing.T) {
 			}
 			mtc.waitForValues(key, []int64{incAB, incAB, incAB})
 
-			// Verify that the cached index and term (Replica.mu.last{Index,Term}))
-			// on all of the replicas is the same. #18327 fixed an issue where the
-			// cached term was left unchanged after applying a snapshot leading to a
-			// persistently unavailable range.
-			repl0, err = mtc.stores[0].GetReplica(1)
-			if err != nil {
-				t.Fatal(err)
-			}
-			expectedLastIndex, _ := repl0.GetLastIndex()
-			expectedLastTerm := repl0.GetCachedLastTerm()
-
-			for i := 1; i < len(mtc.stores); i++ {
-				repl1, err := mtc.stores[i].GetReplica(1)
+			testutils.SucceedsSoon(t, func() error {
+				// Verify that the cached index and term (Replica.mu.last{Index,Term}))
+				// on all of the replicas is the same. #18327 fixed an issue where the
+				// cached term was left unchanged after applying a snapshot leading to a
+				// persistently unavailable range.
+				repl0, err = mtc.stores[0].GetReplica(1)
 				if err != nil {
 					t.Fatal(err)
 				}
-				if lastIndex, _ := repl1.GetLastIndex(); expectedLastIndex != lastIndex {
-					t.Fatalf("%d: expected last index %d, but found %d", i, expectedLastIndex, lastIndex)
+				expectedLastIndex, _ := repl0.GetLastIndex()
+				expectedLastTerm := repl0.GetCachedLastTerm()
+
+				for i := 1; i < len(mtc.stores); i++ {
+					repl1, err := mtc.stores[i].GetReplica(1)
+					if err != nil {
+						return err
+					}
+					if lastIndex, _ := repl1.GetLastIndex(); expectedLastIndex != lastIndex {
+						return fmt.Errorf("%d: expected last index %d, but found %d", i, expectedLastIndex, lastIndex)
+					}
+					if lastTerm := repl1.GetCachedLastTerm(); expectedLastTerm != lastTerm {
+						return fmt.Errorf("%d: expected last term %d, but found %d", i, expectedLastTerm, lastTerm)
+					}
 				}
-				if lastTerm := repl1.GetCachedLastTerm(); expectedLastTerm != lastTerm {
-					t.Fatalf("%d: expected last term %d, but found %d", i, expectedLastTerm, lastTerm)
-				}
-			}
+				return nil
+			})
 		})
 	}
 }

